### PR TITLE
RFC: Add `machinectl import-dkr` command.

### PR DESCRIPTION
**Pull request by richardmaw-codethink**
This adds a `machinectl import-dkr` command, with appropriate hook-up in the systemd-import binary and importd.

I started working on this after discovering that I couldn't pull from a private docker repository, but a separate import would also be useful.

Docker will let you dump a container in two ways:
1.  `docker export` which gives you a flattened tarball, but you need to use it on a running container.
2.  `docker save` which gives you a tarball containing tarballs of the layers, and can be run without running any containers

We can import the result from the first way already with `machinectl import-tar`, but it would be useful to import the result of the second way, as subsequent imports can share the common layers.

I've labelled the pull request RFC because I'm not happy with the amount of code duplication involved, since:
1.  The tarball extraction logic is similar to that used by `import-tar`
2.  The layer extraction logic and state machine is similar to that used by `pull-dkr`.

### CODE REVIEW

**Comment by poettering** [comment]
I tink the general concept of being able to import "docker save" makes sense. I'd be happy if we you'd generalize more of the tarball extraction code into import-common.c or so.

**Comment by poettering** [comment]
I mostly like this, but I figure this patchset should also move import-tar over to use the untarjob stuff..

it is not nice to duplicate the dkr import logic of course, but I think that's something we can live with... might be something to focus more on individual smaller pieces to share between import-dkr and pull-dkr, and accept that the overall logic is not shared... If that makes any sense?

**Comment by richardmaw-codethink** [comment]
On Wed, Aug 05, 2015 at 06:07:30AM -0700, Lennart Poettering wrote:

> I mostly like this, but I figure this patchset should also move import-tar over to use the untarjob stuff..

Yep, I had intended to do that eventually, hence why I put it in a
separate object rather than in-lining the logic.

It turned out to be more work than I had time for before I could put the
Proof of Concept together though, and I'd have had difficulty justifying
work on splitting things out if you wouldn't accept `import-dkr`.

> it is not nice to duplicate the dkr import logic of course,
> but I think that's something we can live with...

I thought about this while working on it,
and I could only come up with ideas which were so abstracted that they'd be hard to follow,
and overkill given there'd only be two or three things that would make use of it.

> might be something to focus more on individual smaller pieces to share between import-dkr and pull-dkr,
> and accept that the overall logic is not shared... If that makes any sense?

Understood.
